challenges/systems/challenge-2.md

```diff
@@ -0,0 +1,294 @@
+opportunity to review your code and provide feedback. Feel free to merge each
+PR after you have two approvals.
+
+Our team will do their best to provide a high quality review of the submitted
+Pull Requests in a reasonable time frame. You are spending your time on this,
+we are going to contribute our time too.
+
+After the final submission, the interview team will assemble and vote using a
+"+1, -2" anonymous voting system: +1 is submitted whenever a team member
+accepts the submission, -2 otherwise.
+
+In case of a positive result, we will connect you to our HR and recruiting
+teams, who will work out the details and present an offer.
+
+In case of a negative score result, hiring manager will contact you and share a
+list of the key observations from the team that affected the result.
+
+# Requirements
+
+## Level 1
+
+### Library
+
+* Implement a simple request forwarder to forward requests to an upstream
+
+### Server
+
+* Develop a simple authorization scheme that defines what upstreams are
+  available to which clients; this scheme can be statically defined in code
+* Secure communication between client and server using TLS and token
+  authentication
+* Accept and forward requests to upstreams using library
+
+## Level 2
+
+### Library
+
+* Implement a round-robin request forwarder to forward requests to
+  upstreams
+
+### Server
+
+* Develop a simple authorization scheme that defines what upstreams are
+  available to which clients; this scheme can be statically defined in code
+* Use mTLS authentication to have the server verify identity of the client and
+  client of the server
+* Accept and forward requests to upstreams using library
+
+## Level 3
+
+### Library
+
+* Track number of connections per upstreams and implement a least connections
+  request forwarder
+
+### Server
+
+* Develop a simple authorization scheme that defines what upstreams are
+  available to which clients; this scheme can be statically defined in code
+* Use mTLS authentication to have the server verify identity of the client and
+  client of the server
+* Accept and forward requests to upstreams using library
+
+## Level 4
+
+### Library
+
+* Track number of connections per upstreams and implement a least connections
+  request forwarder
+* Track number of connections per client and implement a rate limiting request
+  forwarder
+
+### Server
+
+* Develop a simple authorization scheme that defines what upstreams are
+  available to which clients; this scheme can be statically defined in code
+* Use mTLS authentication to have the server verify identity of the client and
+  client of the server
+* Accept and forward requests to upstreams using library
+
+## Level 5
+
+## Library
+
+* Track number of connections per upstreams and implement a least connections
+  request forwarder
+* Track number of connections per client and implement a rate limiting request
+  forwarder
+* Track unhealthy upstreams and implement a circuit breaking requests forwarder
```

Comment:
When I think "circuit breaking" the first thing that comes to mind is client-side circuit breakers. Since we're not asking for a client, should we say something like "implement health checking and remove unhealthy upstreams" instead?